docker/index.js

```diff
@@ -2,100 +2,241 @@ const express = require('express');
-    
-    // pickedOptions = {
-    //     prompt: ...,
-    //     seed: ...,
-    //     guidance_scale: ...,
-    //     num_images_per_prompt: 1
-    // }
-    
-    const SDResult = await axios.get(`${SD_INPAINTING_ENDPOINT}/tasks/${task_id}`, pickedOptions);
-    console.log(JSON.stringify(SDResult.data,null,2));
-
-    if (SDResult.data.status !== "completed") {
-        console.log(`Task ${task_id} is not completed!`);
-        res.send(`Task ${task_id} is not completed!`);
-        return;
-    }
-    
-
-    //pre-check the output path
-    const outputPath = formatPath([...parsedInputPath.slice(0, parsedInputPath.length - 1), "signed_data"]);
-    const result = await ain.db.ref(outputPath).setValue({
-      value: `${JSON.stringify(SDResult.data, null, 2)}`,
-      nonce: -1,
-    })
-    .catch((e) => {
-      console.error(`setValue failure:`, e);
-    });
-  
-});
+	const { transaction } = req.body;
+	const value = transaction.tx_body.operation.value;
+	const taskId = value.task_id;
+	if(cache.get(taskId)){
+		cache.ttl(taskId, 60);
+		return;
+	}
+
+	// if request is first request, set cache 
+	cache.set(taskId, true, 60);
+	
+
+	let uploadMetadataRes; // for catch error
+
+	const inputPath = transaction.tx_body.operation.ref;
+	const parsedInputPath = parsePath(inputPath);
+
+	// pre-check the output path
+	const outputPath = await formatPath([...parsedInputPath.slice(0, parsedInputPath.length - 1), "signed_data"]);
+	const errorPath = formatPath([...parsedInputPath.slice(0, parsedInputPath.length - 1), "error"]);
+
+	console.log(outputPath);
+	// init pinata sdk
+	const pinata = new pinataSDK({ pinataApiKey, pinataSecretApiKey });
+
+
+	// get generated ainft image url with with task id
+	const result = await axios.get(`${process.env.SD_INPAINTING_ENDPOINT}/tasks/${taskId}`);
+
+	const imageUrls = result.data.result;
+
+	// get image file from url
+	const imageDataResponse = await axios.get(imageUrls[value.params.index || 1], {
+		responseType: "arraybuffer",
+	});
+
+	// for test
+	const imageBuffer = fs.readFileSync(__dirname + "/test.png");
+
+	// image file to readable stream
+	const imageDataStream = new Readable();
+	imageDataStream.push(imageBuffer);
+	imageDataStream.push(null);
+
+	// upload image to pinata
+	const options = {
+		pinataMetadata: {
+			name: `${taskId}_image`,
+		},
+		pinataOptions: {
+			cidVersion: 0,
+		},
+	};
+
+	// upload image to ipfs
+	const uploadImgRes = await pinata.pinFileToIPFS(imageDataStream, options)
+		.catch(err => {
+			ain.db.ref(errorPath).setValue({
+				value: {
+					state:"Error",
+					msg:"Image upload fail. check your inforamtion of Image"
+				},
+			})
+			.then(res => console.log(res))
+			.catch((e) => {
+				console.error(`setValue failure:`, e);
+			});
+		});
+
+	// get origin metadata
+	const originMetadata = await axios.get(`https://gateway.pinata.cloud/ipfs/${value.contract_info.metadata_cid}/${value.contract_info.token_id}`, {
```

Comment:
needs error handling ?

Reply:
before handle error, we need to verify code CID and TorkenID. Let's talk about this issue on meeting !